src/target/riscv/riscv-013.c

```diff
@@ -2059,13 +2051,13 @@ static int examine(struct target *target)
 	if (set_dcsr_ebreak(target, false) != ERROR_OK)
 		return ERROR_FAIL;
 
-	target->state = saved_tgt_state;
-	target->debug_reason = saved_dbg_reason;
-
-	if (!halted) {
+	if (state_at_examine_start == RISCV_STATE_RUNNING) {
 		riscv013_step_or_resume_current_hart(target, false);
 		target->state = TARGET_RUNNING;
 		target->debug_reason = DBG_REASON_NOTHALTED;
+	} else if (state_at_examine_start == RISCV_STATE_HALTED) {
+		target->state = TARGET_HALTED;
+		target->debug_reason = DBG_REASON_UNDEFINED;
```

Comment:
Why is this necessary? We know that the hart is halted on line 2054, and the state should reflect that.

Reply:
> Why is this necessary? We know that the hart is halted on line 2054, and the state should reflect that.

Yes, we know that the hart is halted. And we have `target->state = TARGET_HALTED` to indicate that. However, we don't know **why** hart was halted (since we've just connected) and thus we set the reason to UNDEFINED.

Reply:
Got it. Thanks.